Squadrosu.Framework/Piece.cs

```diff
@@ -1,20 +1,169 @@
     /// Direction of the piece
     /// </summary>
-    private int Direction { get; set; }
+    public Direction Direction { get; private set; }
     /// <summary>
     /// Position in the row
     /// </summary>
-    private int Position { get; set; }
+    public int Position { get; private set; }
+    /// <summary>
+    /// Line number
+    /// </summary>
+    public readonly int LineNumber;
+
+    public Piece(Player Player, int LineNumber)
+    {
+        Direction = Direction.Forward;
+        Position = 0;
+        this.LineNumber = LineNumber;
+        this.Player = Player;
+    }
+
+    /// <summary>
+    /// The piece returns to the initial position of the direction
+    /// </summary>
+    public void Reset()
+    {
+        this.Position = (this.Direction == Direction.Forward) ? 0 : 6;
+    }
+    /// <summary>
+    /// Returns the step of a piece in a given state
+    /// </summary>
+    /// <returns>How many squares the piece can go through</returns>
+    public int Step()
+    {
+        if (LineNumber == 3)
+            return 2;
+
+        bool isBlack = Player == Player.Black;
+        bool isForward = Direction == Direction.Forward;
+        bool isOnEdge = LineNumber == 1 || LineNumber == 5;
+        return isBlack ^ isForward ^ isOnEdge ? 3 : 1;
+    }
+
+    /// <summary>
+    /// Move the position of the piece
+    /// </summary>
+    public void Move(Board board)
+    {
+        if (this.Direction == Direction.Forward)
+            this.moveForward(board);
+        else if (this.Direction == Direction.Backward)
+            this.moveBackward(board);
+    }
+
+    /// <summary>
+    /// Move the piece if its direction is forward
+    /// </summary>
+    /// <param name="board">The current board</param>
+    private void moveForward(Board board)
+    {
+        int step = this.Step();
+        int newPosition = Math.Min(this.Position + step, 6);
+        bool isBlack = (this.Player == Player.Black);
+        int i = this.Position;
+
+        while (i + 1 <= Math.Min(this.Position + step, 5))
+        {
+            if (isBlack)
+            {
+                if (board.Positions[this.LineNumber, i + 1] != null)
+                {
+                    while (board.Positions[this.LineNumber, i + 1] != null)
+                    {
+                        board.Positions[this.LineNumber, i + 1].Reset();
+                        newPosition = i + 2;
+                        i++;
+                    }
+                    i = 7;
+                }
+            }
+            else
+            {
+                if (board.Positions[i + 1, this.LineNumber] != null)
+                {
+                    while (board.Positions[i + 1, this.LineNumber] != null)
+                    {
+                        board.Positions[i + 1, this.LineNumber].Reset();
+                        newPosition = i + 2;
+                        i++;
+                    }
+                    i = 7;
+                }
+            }
```

Comment:
Houlà le double `if` nest 7_7
```suggestion
            if (isBlack && board.Positions[LineNumber, i + 1] != null)
            {
                while (board.Positions[LineNumber, i + 1] != null)
                {
                    board.Positions[LineNumber, i + 1].Reset();
                    newPosition = i + 2;
                    i++;
                }
                i = 7;
            }
            else if (board.Positions[i + 1, LineNumber] != null)
            {
                while (board.Positions[i + 1, LineNumber] != null)
                {
                    board.Positions[i + 1, LineNumber].Reset();
                    newPosition = i + 2;
                    i++;
                }
                i = 7;
            }
```

Reply:
Ok